brackend/tasks/auth.py

```diff
@@ -1,46 +1,39 @@
-import datetime
+from functools import wraps
 
-import bcrypt
-import jwt
-from firebase_admin.auth import create_user, generate_email_verification_link
+from firebase_admin.auth import verify_id_token
+from flask import request
+from flask_restful import abort
 
-from brackend.util import SECRET_KEY, BrackendException
 
-EXPIRY_DAYS = 14
+def abrt():
+    """Abort with a 401 response code."""
+    abort(401, message="Authorization Required")
 
 
-class AuthenticationError(BrackendException):
-    pass
+def auth_decorator(meth):
+    """Wraps a restful method with an auth check.
 
+    Stuffs firebase_id into the method.
+    """
 
-def encode_auth_token(user_id):
-    """Generates the Auth Token."""
-    payload = {
-        "exp": datetime.datetime.utcnow() + datetime.timedelta(days=EXPIRY_DAYS),
-        "iat": datetime.datetime.utcnow(),
-        "sub": user_id,
-    }
-    return jwt.encode(payload, SECRET_KEY, algorithm="HS256")
+    @wraps(meth)
+    def wrapper(*args, **kwargs):
+        bearer = request.headers.get("Authorization")
+        if bearer is None:
+            abrt()
+        bearer, token = bearer.split(maxsplit=1)
+        if bearer != "Bearer":
+            abrt()
 
+        # (TODO): Check if user is revoked here
+        try:
+            parsed = verify_id_token(token)
+        except:
+            abrt()
 
-def get_password_and_salt(password):
-    salt = bcrypt.gensalt()
-    password = password.encode()
-    hashed = bcrypt.hashpw(password, salt)
-    return hashed.decode()
+        firebase_id = parsed["uid"]
+        # Stuffs firebase_id into the decorated method
+        # Is there a better way to do this? Parse it from the token again in the method body?
+        return meth(*args, **kwargs, firebase_id=firebase_id)
```

Comment:
We can also do something with the [Flask Global](https://flask.palletsprojects.com/en/2.0.x/api/#flask.g) like:
```
        ...
        from flask import g
        ....
        firebase_id = parsed["uid"]
        user = get_user_by_firebase_id(firebase_id)
        g.user = user
        return meth(*args, **kwargs, firebase_id=firebase_id)
```

Then for any endpoint we can look at `g.user`

Reply:
hmmm, this seems scary to me. I'm worried about persisting state and race conditions. Like what if request A authenticates properly and tries to access some user data, then B comes in and resets g.user before the actual accessing code executes? Would A get B's user data?

Reply:
Nevermind lol
```To share data that is valid for one request only from one function to another, a global variable is not good enough because it would break in threaded environments. Flask provides you with a special object that ensures it is only valid for the active request and that will return different values for each request. In a nutshell: it does the right thing, like it does for request and session.```